python/src/rt_m1_apps/msaf_configuration.py

```diff
@@ -511,6 +568,63 @@ async def sync_configuration(m1: M1Session, streams: dict) -> dict:
                 # The CRC has changed, update it
                 if not await m1.consumptionReportingConfigurationUpdate(ps_id, new_crc):
                     log_error("Failed to update ConsumptionReportingConfiguration for Provisioning Session %s", ps_id)
+
+        # Check for MetricsReportingConfiguration changes in already configured sessions
+        del_metrics = []
+        have_metrics = []
+        new_metrics = List[MetricsReportingConfiguration] = []
+        old_metrics_ids = await m1.metricsReportingConfigurationIds(ps_id)
+        metrics_configs = cfg.get('metricsReporting', None)
+        metrics_list = None
+
+        if metrics_configs is not None:
+            if isinstance(metrics_configs, dict):
+                metrics_list = metrics_configs.items()
+            elif isinstance(metrics_configs, list):
+                metrics_list = [(m.get('samplingPeriod', None), m) for m in metrics_configs]
+            else:
+                log_error(f'Configured metrics for provisioning session "{ps_id}" should be an object or array')
+
+        if old_metrics_ids is None or len(old_metrics_ids) == 0:
+            if metrics_configs is not None:
+                if metrics_list is not None:
+                    for sampling_period, metrics_array in metrics_list:
+                        metricsconfig = dict()
+                        if sampling_period is not None:
+                            metricsconfig.update({'samplingPeriod': sampling_period})
+                        metricsconfig.update(metrics_array)
+                        new_metrics += [metricsconfig]
+        else:
+            new_metrics_left = metrics_list
+            for metrics_id in old_metrics_ids:
+                if new_metrics_left is None or len(new_metrics_left) == 0:
+                    del_metrics += [metrics_id]
+                else:
+                    old_metrics: Optional[MetricsReportingConfiguration] = await m1.metricsReportingConfigurationGet(ps_id, metrics_id)
+                    next_new_metrics_list = []
+                    found = False
+                    for sampling_period, new_metrics in new_metrics_left:
+                        if not found and await metrics_configuration_match(old_metrics, new_metrics):
+                            have_metrics += [metrics_id]
+                            found = True
+                        else:
+                            next_new_metrics_list += [(sampling_period, new_metrics)]
+                    if not found:
+                        del_metrics += [metrics_id]
+                    new_metrics_left = next_new_metrics_list
+
+            for sampling_period, metrics in new_metrics_left:
+                metricsconfig = dict()
+                if sampling_period is not None:
+                    metricsconfig.update({'samplingPeriod': sampling_period})
+                metricsconfig.update(metrics)
+                new_metrics.append(metricsconfig)
```

Comment:
If unique key is dropped then you can simply use the list of left over new metrics instead of having to transform them back into a list. (See previous comments)